// lumentix-complete/frontend/lib/api.ts
import { clearAuthSession, getToken } from "./auth";
import { appPath } from "./paths";

export const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:4000";

type ApiOptions = RequestInit & {
  skipAuth?: boolean;
  token?: string;
};

export async function apiFetch<T>(path: string, options: ApiOptions = {}) {
  const { skipAuth, token: explicitToken, ...fetchOptions } = options;
  const token = explicitToken ?? getToken();
  const headers = new Headers(fetchOptions.headers);

  if (!(options.body instanceof FormData) && !headers.has("Content-Type")) {
    headers.set("Content-Type", "application/json");
  }

  if (!skipAuth && token) {
    headers.set("Authorization", `Bearer ${token}`);
  }

  const response = await fetch(`${API_URL}${path}`, {
    ...fetchOptions,
    headers,
    cache: "no-store"
  });

  if (response.status === 401 && typeof window !== "undefined") {
    clearAuthSession();
    window.location.href = appPath("/login");
  }

  if (!response.ok) {
    const error = await response.json().catch(() => ({ message: "Request failed" }));
    throw new Error(error.message ?? "Request failed");
  }

  if (response.status === 204) {
    return undefined as T;
  }

  return response.json() as Promise<T>;
}

export function backendUrl(path: string) {
  return `${API_URL}${path}`;
}

export async function downloadAuthenticatedFile(path: string, filename: string) {
  const token = getToken();
  const response = await fetch(`${API_URL}${path}`, {
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
    cache: "no-store"
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({ message: "Download failed" }));
    throw new Error(error.message ?? "Download failed");
  }

  const blob = await response.blob();
  const url = window.URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.URL.revokeObjectURL(url);
}
