// lumentix-complete/frontend/lib/types.ts
export type Role = "Admin" | "HR" | "Advocate";

export type InternStatus = "Applied" | "Offered" | "Joined" | "Completed" | "Relieved";

export type ApplicantStatus = "Applied" | "Screening" | "Interview" | "Offered" | "Hired" | "Rejected";

export type DocumentType = "offer" | "certificate" | "experience" | "relieving" | "nda";

export type InternDocument = {
  id: string;
  intern_id: string;
  type: DocumentType;
  file_url: string;
  created_at: string;
};

export type ActivityLog = {
  id: string;
  action: string;
  entity: string;
  entity_id: string | null;
  metadata: Record<string, unknown> | null;
  created_at: string;
};

export type Intern = {
  id: string;
  name: string;
  email: string;
  phone: string;
  role: string;
  department: string;
  joining_date: string;
  exit_date: string | null;
  duration: string;
  manager: string;
  reference_id: string;
  status: InternStatus;
  is_revoked: boolean;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
  documents: InternDocument[];
  activities?: ActivityLog[];
};

export type Applicant = {
  id: string;
  name: string;
  email: string;
  phone: string;
  position: string;
  department: string;
  source: string;
  status: ApplicantStatus;
  experience: string;
  expected_joining_date: string | null;
  resume_url: string | null;
  owner: string;
  notes: string | null;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
};

export type User = {
  id: string;
  email: string;
  role: Role;
  name?: string;
  credit_balance?: number;
  free_case_expires_at?: string | null;
  free_case_used?: boolean;
};

export type AdvocateCase = {
  id: string;
  case_title: string;
  client_name: string;
  case_type: string;
  jurisdiction: string | null;
  description: string;
  status: "Open" | "In Review" | "Closed";
  credit_source: "Free Trial" | "Paid Credit";
  created_at: string;
  updated_at: string;
};

export type CreditTransaction = {
  id: string;
  type: string;
  credits: number;
  status: "Completed" | "Pending Approval" | "Rejected";
  package_id: string | null;
  amount_inr: number | null;
  case_id: string | null;
  metadata: Record<string, unknown> | null;
  created_at: string;
};

export type AdvocateDashboard = {
  advocate: {
    id: string;
    name: string;
    email: string;
    phone: string | null;
    organization: string | null;
    bar_council_id: string | null;
  };
  credits: {
    credit_balance: number;
    free_case_available: boolean;
    free_case_used: boolean;
    free_case_expires_at: string | null;
  };
  cases: AdvocateCase[];
  transactions: CreditTransaction[];
};

export type VerificationRecord = {
  verification_id: string;
  verified_at: string;
  reference_id: string;
  status: "Valid" | "Revoked";
  lifecycle_status: InternStatus;
  issue_date: string;
  joining_date: string;
  exit_date: string;
  name: string;
  role: string;
  department: string;
  duration: string;
  reporting_manager: string;
  registered_email: string;
  registered_phone: string;
  company: string;
  company_address: string;
  bgv_email: string;
  record_scope: string;
  documents_issued: Array<{
    type: DocumentType;
    label: string;
    issued_at: string;
  }>;
  verifier: {
    name: string;
    email: string;
    organization: string;
    designation: string | null;
    purpose: string;
  };
};

export type BgvAccessLog = {
  id: string;
  action: "PUBLIC_VERIFICATION_COMPLETED" | "PUBLIC_VERIFICATION_SUMMARY_DOWNLOADED";
  access_type: string;
  created_at: string;
  reference_id: string;
  result_status: "Valid" | "Revoked";
  intern: {
    id: string;
    name: string;
    reference_id: string;
  } | null;
  verifier: {
    name: string;
    email: string;
    phone: string;
    organization: string;
    designation: string;
    purpose: string;
    notes: string;
  };
  request: {
    ip_address: string;
    user_agent: string;
  };
};

export type SystemStatus = {
  service: string;
  generated_at: string;
  environment: string;
  database: {
    driver: string;
    project_id: string;
    emulator_host: string | null;
    health: Record<string, unknown>;
    collections: string[];
  };
  storage: {
    driver: string;
    local_directory: string | null;
  };
  verification: {
    public_verify_base_url: string;
    frontend_origins: string[];
  };
  email: {
    smtp_configured: boolean;
    host: string | null;
    port: number;
    user: string | null;
    from: string;
  };
  readiness: Array<{
    key: string;
    label: string;
    status: "ok" | "attention";
    detail: string;
  }>;
};
