// lumentix-complete/frontend/components/applicants/applicant-form.tsx
"use client";

import { useState } from "react";
import { useRouter } from "next/navigation";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { Select } from "@/components/ui/select";
import { Textarea } from "@/components/ui/textarea";
import { apiFetch } from "@/lib/api";
import type { Applicant, ApplicantStatus } from "@/lib/types";

const statuses: ApplicantStatus[] = ["Applied", "Screening", "Interview", "Offered", "Hired", "Rejected"];

type FormState = {
  name: string;
  email: string;
  phone: string;
  position: string;
  department: string;
  source: string;
  status: ApplicantStatus;
  experience: string;
  expected_joining_date: string;
  resume_url: string;
  owner: string;
  notes: string;
};

const emptyState: FormState = {
  name: "",
  email: "",
  phone: "",
  position: "",
  department: "",
  source: "",
  status: "Applied",
  experience: "",
  expected_joining_date: "",
  resume_url: "",
  owner: "",
  notes: ""
};

function experienceYears(value: string) {
  if (/^fresher$/i.test(value.trim())) return "0";
  return value.match(/\d{1,2}(\.\d{1,2})?/)?.[0] ?? "";
}

function experienceLabel(value: string) {
  const normalized = Number(value);
  if (!Number.isFinite(normalized) || normalized < 0) return "";
  if (normalized === 0) return "Fresher";
  return `${value} ${normalized === 1 ? "year" : "years"}`;
}

export function ApplicantForm({ initial }: { initial?: Applicant }) {
  const router = useRouter();
  const [form, setForm] = useState<FormState>(
    initial
      ? {
          name: initial.name,
          email: initial.email,
          phone: initial.phone,
          position: initial.position,
          department: initial.department,
          source: initial.source,
          status: initial.status,
          experience: experienceYears(initial.experience),
          expected_joining_date: initial.expected_joining_date ? initial.expected_joining_date.slice(0, 10) : "",
          resume_url: initial.resume_url ?? "",
          owner: initial.owner,
          notes: initial.notes ?? ""
        }
      : emptyState
  );
  const [saving, setSaving] = useState(false);
  const [error, setError] = useState("");

  function update<K extends keyof FormState>(key: K, value: FormState[K]) {
    setForm((current) => {
      let normalizedValue = value;

      if (key === "phone") {
        normalizedValue = String(value).replace(/\D/g, "") as FormState[K];
      }

      if (key === "experience") {
        normalizedValue = String(value).replace(/[^\d.]/g, "").replace(/(\..*)\./g, "$1") as FormState[K];
      }

      return { ...current, [key]: normalizedValue };
    });
  }

  async function submit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    setSaving(true);
    setError("");

    try {
      const payload = {
        ...form,
        experience: experienceLabel(form.experience)
      };

      if (initial) {
        await apiFetch(`/api/applicant/${initial.id}`, {
          method: "PUT",
          body: JSON.stringify(payload)
        });
        router.push("/applicants");
        router.refresh();
      } else {
        const response = await apiFetch<{ data: Applicant }>("/api/applicant", {
          method: "POST",
          body: JSON.stringify(payload)
        });
        router.push(`/applicants/${response.data.id}`);
      }
    } catch (err) {
      setError(err instanceof Error ? err.message : "Unable to save applicant");
    } finally {
      setSaving(false);
    }
  }

  return (
    <form onSubmit={submit} className="rounded-lg border border-border bg-background p-5 shadow-soft">
      <div className="grid gap-4 md:grid-cols-2">
        <Field label="Name">
          <Input value={form.name} onChange={(event) => update("name", event.target.value)} pattern="[A-Za-z .'-]{2,120}" title="Use letters only." required />
        </Field>
        <Field label="Email">
          <Input type="email" value={form.email} onChange={(event) => update("email", event.target.value)} required />
        </Field>
        <Field label="Phone">
          <Input value={form.phone} onChange={(event) => update("phone", event.target.value)} inputMode="numeric" pattern="[0-9]{7,15}" title="Use digits only, 7 to 15 numbers." required />
        </Field>
        <Field label="Position">
          <Input value={form.position} onChange={(event) => update("position", event.target.value)} minLength={2} maxLength={120} required />
        </Field>
        <Field label="Department">
          <Input value={form.department} onChange={(event) => update("department", event.target.value)} minLength={2} maxLength={120} required />
        </Field>
        <Field label="Source">
          <Input placeholder="Referral, LinkedIn, careers page" value={form.source} onChange={(event) => update("source", event.target.value)} minLength={2} maxLength={120} required />
        </Field>
        <Field label="Experience (Years)">
          <Input type="number" min="0" max="60" step="0.1" placeholder="0 for fresher" value={form.experience} onChange={(event) => update("experience", event.target.value)} required />
        </Field>
        <Field label="Expected Joining Date">
          <Input type="date" value={form.expected_joining_date} onChange={(event) => update("expected_joining_date", event.target.value)} />
        </Field>
        <Field label="Recruiter / Owner">
          <Input value={form.owner} onChange={(event) => update("owner", event.target.value)} pattern="[A-Za-z .'-]{2,120}" title="Use letters only." required />
        </Field>
        <Field label="Pipeline Stage">
          <Select value={form.status} onChange={(event) => update("status", event.target.value as ApplicantStatus)}>
            {statuses.map((status) => (
              <option key={status} value={status}>
                {status}
              </option>
            ))}
          </Select>
        </Field>
        <Field label="Resume Link">
          <Input type="url" value={form.resume_url} onChange={(event) => update("resume_url", event.target.value)} />
        </Field>
        <div className="space-y-2 md:col-span-2">
          <Label>Notes</Label>
          <Textarea value={form.notes} onChange={(event) => update("notes", event.target.value)} />
        </div>
      </div>
      {error ? <p className="mt-4 text-sm text-destructive">{error}</p> : null}
      <div className="mt-6 flex justify-end gap-2">
        <Button type="button" variant="outline" onClick={() => router.back()}>
          Cancel
        </Button>
        <Button type="submit" disabled={saving}>
          {saving ? "Saving..." : initial ? "Save Applicant" : "Add Applicant"}
        </Button>
      </div>
    </form>
  );
}

function Field({ label, children }: { label: string; children: React.ReactNode }) {
  return (
    <div className="space-y-2">
      <Label>{label}</Label>
      {children}
    </div>
  );
}
