// lumentix-complete/frontend/app/(protected)/dashboard/page.tsx
"use client";

import Link from "next/link";
import { useEffect, useMemo, useState } from "react";
import { Activity, FileCheck2, FileText, ShieldCheck, UserRoundPlus, UserSearch, Users } from "lucide-react";
import { ApplicantStatusBadge } from "@/components/applicants/applicant-status-badge";
import { Card, CardContent, CardHeader, CardTitle } from "@/components/ui/card";
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from "@/components/ui/table";
import { StatusBadge } from "@/components/status-badge";
import { apiFetch } from "@/lib/api";
import type { Applicant, ApplicantStatus, Intern, InternStatus } from "@/lib/types";

const lifecycle: InternStatus[] = ["Applied", "Offered", "Joined", "Completed", "Relieved"];
const applicantLifecycle: ApplicantStatus[] = ["Applied", "Screening", "Interview", "Offered", "Hired", "Rejected"];

type CountMeta = {
  total: number;
};

export default function DashboardPage() {
  const [interns, setInterns] = useState<Intern[]>([]);
  const [applicants, setApplicants] = useState<Applicant[]>([]);
  const [bgvAccessTotal, setBgvAccessTotal] = useState(0);
  const [generatedDocumentTotal, setGeneratedDocumentTotal] = useState(0);
  const [loading, setLoading] = useState(true);
  const [message, setMessage] = useState("");

  useEffect(() => {
    Promise.all([
      apiFetch<{ data: Intern[] }>("/api/interns?limit=100"),
      apiFetch<{ data: Applicant[] }>("/api/applicants?limit=100"),
      apiFetch<{ data: unknown[]; meta: CountMeta }>("/api/bgv/access-logs?page=1&limit=1"),
      apiFetch<{ data: unknown[]; meta: CountMeta }>("/api/documents?offset=0&limit=1")
    ])
      .then(([internResponse, applicantResponse, bgvResponse, documentResponse]) => {
        setInterns(internResponse.data);
        setApplicants(applicantResponse.data);
        setBgvAccessTotal(bgvResponse.meta.total);
        setGeneratedDocumentTotal(documentResponse.meta.total);
      })
      .catch((err) => {
        setMessage(err instanceof Error ? err.message : "Unable to load dashboard metrics");
      })
      .finally(() => setLoading(false));
  }, []);

  const stats = useMemo(() => {
    const byStatus = new Map<InternStatus, number>();
    lifecycle.forEach((status) => byStatus.set(status, 0));
    interns.forEach((intern) => byStatus.set(intern.status, (byStatus.get(intern.status) ?? 0) + 1));
    return {
      total: interns.length,
      active: interns.filter((intern) => intern.status === "Joined").length,
      documents: interns.reduce((sum, intern) => sum + intern.documents.length, 0),
      completed: interns.filter((intern) => intern.status === "Completed" || intern.status === "Relieved").length,
      byStatus
    };
  }, [interns]);

  const applicantStats = useMemo(() => {
    const byStatus = new Map<ApplicantStatus, number>();
    applicantLifecycle.forEach((status) => byStatus.set(status, 0));
    applicants.forEach((applicant) => byStatus.set(applicant.status, (byStatus.get(applicant.status) ?? 0) + 1));
    return {
      total: applicants.length,
      active: applicants.filter((applicant) => !["Hired", "Rejected"].includes(applicant.status)).length,
      interviews: applicants.filter((applicant) => applicant.status === "Interview").length,
      offers: applicants.filter((applicant) => applicant.status === "Offered").length,
      byStatus
    };
  }, [applicants]);

  return (
    <div className="space-y-6">
      <div className="flex flex-col gap-3 md:flex-row md:items-end md:justify-between">
        <div>
          <p className="text-sm uppercase text-muted-foreground">Admin dashboard</p>
          <h1 className="text-3xl font-semibold">Employment lifecycle control center</h1>
        </div>
        <div className="flex flex-wrap gap-2">
          <Link className="inline-flex h-10 items-center justify-center rounded-lg border border-border bg-background px-4 text-sm font-medium hover:bg-muted" href="/bgv/access">
            <ShieldCheck className="mr-2 h-4 w-4" />
            BGV Access
          </Link>
          <Link className="inline-flex h-10 items-center justify-center rounded-lg border border-border bg-background px-4 text-sm font-medium hover:bg-muted" href="/system/status">
            <Activity className="mr-2 h-4 w-4" />
            System Status
          </Link>
          <Link className="inline-flex h-10 items-center justify-center rounded-lg border border-border bg-background px-4 text-sm font-medium hover:bg-muted" href="/applicants/add">
            <UserSearch className="mr-2 h-4 w-4" />
            Add Applicant
          </Link>
          <Link className="inline-flex h-10 items-center justify-center rounded-lg bg-primary px-4 text-sm font-medium text-primary-foreground hover:bg-primary/90" href="/interns/add">
            <UserRoundPlus className="mr-2 h-4 w-4" />
            Add Record
          </Link>
        </div>
      </div>

      {message ? <p className="rounded-lg border border-red-300 bg-red-50 p-3 text-sm text-red-900 dark:border-red-500/40 dark:bg-red-500/15 dark:text-red-200">{message}</p> : null}

      <div className="grid gap-4 md:grid-cols-4">
        <Metric title="Total Records" value={stats.total} icon={Users} />
        <Metric title="Joined" value={stats.active} icon={ShieldCheck} />
        <Metric title="Linked Documents" value={stats.documents} icon={FileCheck2} />
        <Metric title="Completed" value={stats.completed} icon={FileCheck2} />
      </div>

      <div className="grid gap-4 md:grid-cols-4">
        <Metric title="Applicants" value={applicantStats.total} icon={UserSearch} />
        <Metric title="Active Pipeline" value={applicantStats.active} icon={Users} />
        <Metric title="Interviews" value={applicantStats.interviews} icon={UserRoundPlus} />
        <Metric title="Offers" value={applicantStats.offers} icon={ShieldCheck} />
      </div>

      <div className="grid gap-4 md:grid-cols-4">
        <Metric title="BGV Access Logs" value={bgvAccessTotal} icon={ShieldCheck} />
        <Metric title="Generated Files" value={generatedDocumentTotal} icon={FileText} />
        <Metric title="Revoked BGV" value={interns.filter((intern) => intern.is_revoked).length} icon={Activity} />
        <Metric title="Open Records" value={interns.filter((intern) => !["Completed", "Relieved"].includes(intern.status)).length} icon={Users} />
      </div>

      <div className="rounded-lg border border-border bg-background p-5 shadow-soft">
        <h2 className="mb-4 text-base font-semibold">Lifecycle Pipeline</h2>
        <div className="grid gap-3 md:grid-cols-5">
          {lifecycle.map((status) => (
            <div key={status} className="rounded-lg border border-border p-4">
              <div className="mb-3">
                <StatusBadge status={status} />
              </div>
              <div className="text-2xl font-semibold">{stats.byStatus.get(status)}</div>
            </div>
          ))}
        </div>
      </div>

      <div className="rounded-lg border border-border bg-background p-5 shadow-soft">
        <h2 className="mb-4 text-base font-semibold">Applicant Pipeline</h2>
        <div className="grid gap-3 md:grid-cols-6">
          {applicantLifecycle.map((status) => (
            <div key={status} className="rounded-lg border border-border p-4">
              <div className="mb-3">
                <ApplicantStatusBadge status={status} />
              </div>
              <div className="text-2xl font-semibold">{applicantStats.byStatus.get(status)}</div>
            </div>
          ))}
        </div>
      </div>

      <div className="rounded-lg border border-border bg-background shadow-soft">
        <div className="border-b border-border p-5">
          <h2 className="text-base font-semibold">Recent Records</h2>
        </div>
        <Table>
          <TableHeader>
            <TableRow>
              <TableHead>Name</TableHead>
              <TableHead>Reference</TableHead>
              <TableHead>Status</TableHead>
              <TableHead className="text-right">Open</TableHead>
            </TableRow>
          </TableHeader>
          <TableBody>
            {loading ? (
              <TableRow>
                <TableCell colSpan={4} className="text-center text-muted-foreground">
                  Loading...
                </TableCell>
              </TableRow>
            ) : (
              interns.slice(0, 8).map((intern) => (
                <TableRow key={intern.id}>
                  <TableCell>
                    <div className="font-medium">{intern.name}</div>
                    <div className="text-xs text-muted-foreground">{intern.role}</div>
                  </TableCell>
                  <TableCell className="font-mono text-xs">{intern.reference_id}</TableCell>
                  <TableCell>
                    <StatusBadge status={intern.status} />
                  </TableCell>
                  <TableCell className="text-right">
                    <Link className="text-sm font-medium text-primary hover:underline" href={`/interns/${intern.id}`}>
                      View
                    </Link>
                  </TableCell>
                </TableRow>
              ))
            )}
          </TableBody>
        </Table>
      </div>
    </div>
  );
}

function Metric({
  title,
  value,
  icon: Icon
}: {
  title: string;
  value: number;
  icon: React.ComponentType<{ className?: string }>;
}) {
  return (
    <Card>
      <CardHeader className="flex flex-row items-center justify-between pb-2">
        <CardTitle className="text-sm text-muted-foreground">{title}</CardTitle>
        <Icon className="h-4 w-4 text-primary" />
      </CardHeader>
      <CardContent>
        <div className="text-3xl font-semibold">{value}</div>
      </CardContent>
    </Card>
  );
}
