// lumentix-complete/frontend/app/verify/[reference_id]/verification-client.tsx
"use client";

import Image from "next/image";
import { useState } from "react";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { Select } from "@/components/ui/select";
import { Textarea } from "@/components/ui/textarea";
import { API_URL, apiFetch } from "@/lib/api";
import { publicAssetPath } from "@/lib/paths";
import type { VerificationRecord } from "@/lib/types";

const purposes = [
  "Background Verification",
  "Employment Verification",
  "Academic Verification",
  "Client Verification",
  "Vendor Verification",
  "Other"
];

type FormState = {
  verifier_name: string;
  verifier_email: string;
  verifier_phone: string;
  verifier_organization: string;
  verifier_designation: string;
  purpose: string;
  notes: string;
  consent: boolean;
};

const initialForm: FormState = {
  verifier_name: "",
  verifier_email: "",
  verifier_phone: "",
  verifier_organization: "",
  verifier_designation: "",
  purpose: "Background Verification",
  notes: "",
  consent: false
};

export function VerificationClient({ referenceId }: { referenceId: string }) {
  const [form, setForm] = useState<FormState>(initialForm);
  const [record, setRecord] = useState<VerificationRecord | null>(null);
  const [verifiedForm, setVerifiedForm] = useState<FormState | null>(null);
  const [error, setError] = useState("");
  const [downloadError, setDownloadError] = useState("");
  const [loading, setLoading] = useState(false);
  const [downloading, setDownloading] = useState(false);

  function update<K extends keyof FormState>(key: K, value: FormState[K]) {
    setForm((current) => {
      const next = { ...current, [key]: value };

      if (key === "verifier_phone") {
        next[key] = String(value).replace(/\D/g, "") as FormState[K];
      }

      return next;
    });
  }

  async function submit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    setError("");
    setLoading(true);

    try {
      const response = await apiFetch<{ data: VerificationRecord }>(`/api/verify/${referenceId}`, {
        method: "POST",
        skipAuth: true,
        body: JSON.stringify(form)
      });
      setRecord(response.data);
      setVerifiedForm({ ...form });
      setDownloadError("");
    } catch (err) {
      setError(err instanceof Error ? err.message : "Unable to verify this reference");
    } finally {
      setLoading(false);
    }
  }

  async function downloadSummary() {
    const downloadForm = verifiedForm ?? form;

    if (!record) return;
    setDownloading(true);
    setDownloadError("");

    try {
      const response = await fetch(`${API_URL}/api/verify/${referenceId}/download`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(downloadForm)
      });

      if (!response.ok) {
        const errorBody = await response.json().catch(() => ({ message: "Download failed" }));
        throw new Error(errorBody.message ?? "Download failed");
      }

      const blob = await response.blob();
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement("a");
      const safeReference = record.reference_id.replace(/[^a-zA-Z0-9_-]/g, "-");
      link.href = url;
      link.download = `${safeReference}-verification-summary.pdf`;
      document.body.appendChild(link);
      link.click();
      link.remove();
      window.URL.revokeObjectURL(url);
    } catch (err) {
      setDownloadError(err instanceof Error ? err.message : "Unable to download verification summary");
    } finally {
      setDownloading(false);
    }
  }

  return (
    <main className="min-h-screen bg-muted/30">
      <section className="mx-auto flex min-h-screen max-w-5xl flex-col justify-center p-4">
        <div className="rounded-lg border border-border bg-background p-6 shadow-soft md:p-8">
          <Image src={publicAssetPath("/lumentix-logo.png")} alt="Lumentix" width={240} height={90} priority />
          <div className="mt-8 max-w-3xl">
            <p className="text-sm uppercase text-muted-foreground">Controlled document verification</p>
            <h1 className="mt-2 text-3xl font-semibold">LUMENTIX HRFLOW</h1>
            <p className="mt-3 text-sm text-muted-foreground">
              Verification details are released only after verifier information is submitted. Each
              request is recorded for audit and background verification tracking.
            </p>
          </div>

          <div className="mt-8 grid gap-6 lg:grid-cols-[380px_1fr]">
            <form onSubmit={submit} className="space-y-4 rounded-lg border border-border p-4">
              <div>
                <p className="text-xs uppercase text-muted-foreground">Reference number</p>
                <p className="font-mono text-sm font-medium">{referenceId}</p>
              </div>
              <Field id="verifier-name" label="Verifier Name">
                <Input id="verifier-name" value={form.verifier_name} onChange={(event) => update("verifier_name", event.target.value)} pattern="[A-Za-z .'-]{2,120}" title="Use letters only." required />
              </Field>
              <Field id="verifier-email" label="Official Email">
                <Input id="verifier-email" type="email" value={form.verifier_email} onChange={(event) => update("verifier_email", event.target.value)} required />
              </Field>
              <Field id="verifier-phone" label="Phone">
                <Input id="verifier-phone" value={form.verifier_phone} onChange={(event) => update("verifier_phone", event.target.value)} inputMode="numeric" pattern="[0-9]{7,15}" placeholder="Optional" />
              </Field>
              <Field id="verifier-organization" label="Company / Institution">
                <Input id="verifier-organization" value={form.verifier_organization} onChange={(event) => update("verifier_organization", event.target.value)} required />
              </Field>
              <Field id="verifier-designation" label="Designation">
                <Input id="verifier-designation" value={form.verifier_designation} onChange={(event) => update("verifier_designation", event.target.value)} placeholder="Optional" />
              </Field>
              <Field id="verification-purpose" label="Verification Purpose">
                <Select id="verification-purpose" value={form.purpose} onChange={(event) => update("purpose", event.target.value)}>
                  {purposes.map((purpose) => (
                    <option key={purpose} value={purpose}>
                      {purpose}
                    </option>
                  ))}
                </Select>
              </Field>
              <Field id="verification-notes" label="Notes">
                <Textarea id="verification-notes" value={form.notes} onChange={(event) => update("notes", event.target.value)} placeholder="Optional context or case ID" rows={3} />
              </Field>
              <label className="flex gap-3 text-sm text-muted-foreground">
                <input
                  className="mt-1 h-4 w-4"
                  type="checkbox"
                  checked={form.consent}
                  onChange={(event) => update("consent", event.target.checked)}
                  required
                />
                <span>I confirm that this request is for a legitimate verification purpose and may be recorded by LUMENTIX.</span>
              </label>
              {error ? <p className="rounded-lg border border-red-300 bg-red-50 p-3 text-sm text-red-900 dark:border-red-500/40 dark:bg-red-500/15 dark:text-red-200">{error}</p> : null}
              <Button className="w-full" disabled={loading}>
                {loading ? "Preparing Report..." : "Get Verification Report"}
              </Button>
            </form>

            <div className="rounded-lg border border-border p-4">
              {!record ? (
                <div className="flex h-full min-h-[360px] flex-col justify-center">
                  <p className="text-sm uppercase text-muted-foreground">Access controlled</p>
                  <h2 className="mt-2 text-2xl font-semibold">Submit verifier details to view the record.</h2>
                  <p className="mt-3 text-sm text-muted-foreground">
                    The result will include role, department, joining date, lifecycle status,
                    issued document summary, and company BGV contact details.
                  </p>
                </div>
              ) : (
                <VerificationResult
                  record={record}
                  downloadError={downloadError}
                  downloading={downloading}
                  onDownload={downloadSummary}
                />
              )}
            </div>
          </div>
        </div>
      </section>
    </main>
  );
}

function VerificationResult({
  record,
  downloadError,
  downloading,
  onDownload
}: {
  record: VerificationRecord;
  downloadError: string;
  downloading: boolean;
  onDownload: () => void;
}) {
  return (
    <div className="space-y-6">
      <div className="flex items-start justify-between gap-3 border-b border-border pb-5">
        <div>
          <p className="text-xs uppercase text-muted-foreground">Verified record</p>
          <h2 className="mt-1 text-2xl font-semibold">{record.name}</h2>
          <p className="font-mono text-sm text-muted-foreground">{record.reference_id}</p>
        </div>
        <div className="flex flex-col items-end gap-3">
          <Badge variant={record.status === "Valid" ? "teal" : "red"}>{record.status}</Badge>
          <Button type="button" variant="outline" size="sm" onClick={onDownload} disabled={downloading}>
            {downloading ? "Preparing..." : "Download One-Page Summary"}
          </Button>
        </div>
      </div>
      {downloadError ? (
        <p className="rounded-lg border border-red-300 bg-red-50 p-3 text-sm text-red-900 dark:border-red-500/40 dark:bg-red-500/15 dark:text-red-200">
          {downloadError}
        </p>
      ) : null}

      <dl className="grid gap-4 md:grid-cols-2">
        <Item label="Role" value={record.role} />
        <Item label="Department" value={record.department} />
        <Item label="Joining Date" value={record.joining_date} />
        <Item label="Date Of Exit" value={record.exit_date} />
        <Item label="Duration" value={record.duration} />
        <Item label="Lifecycle Status" value={record.lifecycle_status} />
        <Item label="Issue Date" value={record.issue_date} />
        <Item label="Reporting Manager" value={record.reporting_manager} />
        <Item label="Registered Email" value={record.registered_email} />
        <Item label="Registered Phone" value={record.registered_phone} />
        <Item label="BGV Email" value={record.bgv_email} />
      </dl>

      <div className="rounded-lg border border-border bg-muted/30 p-4">
        <p className="text-xs uppercase text-muted-foreground">Company</p>
        <p className="mt-1 text-sm font-semibold">{record.company}</p>
        <p className="mt-1 text-sm text-muted-foreground">{record.company_address}</p>
        <p className="mt-3 text-sm text-muted-foreground">{record.record_scope}</p>
      </div>

      <div>
        <h3 className="text-sm font-semibold">Issued Documents</h3>
        {record.documents_issued.length === 0 ? (
          <p className="mt-2 text-sm text-muted-foreground">No documents are recorded for this reference yet.</p>
        ) : (
          <div className="mt-3 grid gap-2">
            {record.documents_issued.map((document, index) => (
              <div key={`${document.type}-${index}`} className="flex items-center justify-between rounded-lg border border-border p-3 text-sm">
                <span>{document.label}</span>
                <span className="text-muted-foreground">{document.issued_at}</span>
              </div>
            ))}
          </div>
        )}
      </div>

      <div className="rounded-lg border border-border p-4">
        <p className="text-xs uppercase text-muted-foreground">Verifier logged</p>
        <p className="mt-1 text-sm font-medium">{record.verifier.name}</p>
        <p className="text-sm text-muted-foreground">
          {record.verifier.organization} | {record.verifier.email}
        </p>
        <p className="mt-2 text-xs text-muted-foreground">
          Verification ID: {record.verification_id} | {new Date(record.verified_at).toLocaleString()}
        </p>
      </div>
    </div>
  );
}

function Field({ id, label, children }: { id: string; label: string; children: React.ReactNode }) {
  return (
    <div className="space-y-2">
      <Label htmlFor={id}>{label}</Label>
      {children}
    </div>
  );
}

function Item({ label, value }: { label: string; value: string }) {
  return (
    <div>
      <dt className="text-xs uppercase text-muted-foreground">{label}</dt>
      <dd className="mt-1 text-sm font-medium">{value}</dd>
    </div>
  );
}
