// lumentix-complete/frontend/components/site/AdvancedGraphics.tsx
"use client";

import Image from "next/image";
import Link from "next/link";
import { motion } from "framer-motion";
import { ArrowRight, BrainCircuit, Cpu, Database, GitBranch, LockKeyhole, Network, Sparkles, Workflow } from "lucide-react";
import { cn } from "@/lib/utils";

export type FeatureItem = {
  title: string;
  description: string;
  meta?: string;
};

export type StatItem = {
  value: string;
  label: string;
};

export function Reveal({
  children,
  className = "",
  delay = 0,
  direction = "up"
}: {
  children: React.ReactNode;
  className?: string;
  delay?: number;
  direction?: "up" | "left" | "right" | "none";
}) {
  const offset = {
    up: { x: 0, y: 28 },
    left: { x: -32, y: 0 },
    right: { x: 32, y: 0 },
    none: { x: 0, y: 0 }
  }[direction];

  return (
    <motion.div
      initial={{ opacity: 0, ...offset }}
      whileInView={{ opacity: 1, x: 0, y: 0 }}
      viewport={{ once: true, margin: "-80px" }}
      transition={{ duration: 0.55, delay }}
      className={className}
    >
      {children}
    </motion.div>
  );
}

const fadeUp = {
  hidden: { opacity: 0, y: 28 },
  visible: { opacity: 1, y: 0 }
};

const stagger = {
  hidden: {},
  visible: {
    transition: {
      staggerChildren: 0.1
    }
  }
};

export function Eyebrow({ children }: { children: React.ReactNode }) {
  return (
    <motion.span
      initial={{ opacity: 0, y: 10 }}
      whileInView={{ opacity: 1, y: 0 }}
      viewport={{ once: true, margin: "-80px" }}
      transition={{ duration: 0.45 }}
      className="inline-flex items-center gap-2 border border-primary/20 bg-primary/10 px-3 py-1 text-xs font-semibold uppercase tracking-[0.22em] text-primary"
    >
      <Sparkles className="h-3.5 w-3.5 animate-pulse" />
      {children}
    </motion.span>
  );
}

export function SectionHeader({
  eyebrow,
  title,
  description,
  align = "center"
}: {
  eyebrow: string;
  title: string;
  description: string;
  align?: "center" | "left";
}) {
  return (
    <motion.div
      initial="hidden"
      whileInView="visible"
      viewport={{ once: true, margin: "-90px" }}
      variants={stagger}
      className={cn("space-y-4", align === "center" ? "mx-auto max-w-3xl text-center" : "max-w-3xl")}
    >
      <Eyebrow>{eyebrow}</Eyebrow>
      <motion.h2 variants={fadeUp} transition={{ duration: 0.55 }} className="text-balance text-3xl font-semibold tracking-tight text-foreground md:text-5xl">
        {title}
      </motion.h2>
      <motion.p variants={fadeUp} transition={{ duration: 0.55 }} className="text-base leading-8 text-muted-foreground md:text-lg">
        {description}
      </motion.p>
    </motion.div>
  );
}

export function AINetworkGraphic({ className = "" }: { className?: string }) {
  const nodes = [
    { label: "Ideas", x: "12%", y: "22%" },
    { label: "Design", x: "38%", y: "14%" },
    { label: "AI", x: "69%", y: "25%" },
    { label: "Products", x: "18%", y: "67%" },
    { label: "Brand", x: "51%", y: "58%" },
    { label: "Growth", x: "78%", y: "72%" }
  ];

  return (
    <motion.div
      initial={{ opacity: 0, scale: 0.96, y: 24 }}
      whileInView={{ opacity: 1, scale: 1, y: 0 }}
      viewport={{ once: true, margin: "-80px" }}
      transition={{ duration: 0.7 }}
      className={cn("relative min-h-[420px] overflow-hidden border border-border bg-white shadow-elegant", className)}
    >
      <div className="absolute inset-0 site-grid opacity-80" />
      <div className="absolute inset-x-0 top-0 h-1 bg-[linear-gradient(90deg,hsl(var(--primary)),hsl(var(--accent)),hsl(var(--signal)))]" />
      <motion.div
        animate={{ rotate: 360 }}
        transition={{ duration: 28, repeat: Infinity, ease: "linear" }}
        className="absolute left-1/2 top-1/2 h-52 w-52 -translate-x-1/2 -translate-y-1/2 border border-primary/25 bg-primary/10"
      />
      <motion.div
        animate={{ rotate: -360 }}
        transition={{ duration: 20, repeat: Infinity, ease: "linear" }}
        className="absolute left-1/2 top-1/2 h-32 w-32 -translate-x-1/2 -translate-y-1/2 border border-accent/40 bg-accent/10"
      />
      <motion.div
        animate={{ y: [0, -8, 0] }}
        transition={{ duration: 3.4, repeat: Infinity, ease: "easeInOut" }}
        className="absolute left-1/2 top-1/2 grid h-20 w-20 -translate-x-1/2 -translate-y-1/2 place-items-center border border-primary/20 bg-primary text-primary-foreground"
      >
        <BrainCircuit className="h-9 w-9" />
      </motion.div>
      <svg className="absolute inset-0 h-full w-full" aria-hidden="true">
        {[
          ["12%", "22%", "text-primary/30"],
          ["38%", "14%", "text-accent/40"],
          ["69%", "25%", "text-primary/30"],
          ["18%", "67%", "text-signal/40"],
          ["51%", "58%", "text-primary/30"],
          ["78%", "72%", "text-accent/40"]
        ].map(([x, y, color], index) => (
          <motion.line
            key={`${x}-${y}`}
            x1="50%"
            y1="50%"
            x2={x}
            y2={y}
            stroke="currentColor"
            className={color}
            strokeWidth="1.5"
            initial={{ pathLength: 0, opacity: 0 }}
            whileInView={{ pathLength: 1, opacity: 1 }}
            viewport={{ once: true }}
            transition={{ duration: 0.8, delay: index * 0.08 }}
          />
        ))}
      </svg>
      {nodes.map((node, index) => (
        <motion.div
          key={node.label}
          className="absolute -translate-x-1/2 -translate-y-1/2 border border-border bg-background px-3 py-2 text-xs font-semibold text-foreground shadow-elegant"
          style={{ left: node.x, top: node.y }}
          initial={{ opacity: 0, scale: 0.7 }}
          whileInView={{ opacity: 1, scale: 1 }}
          viewport={{ once: true }}
          transition={{ duration: 0.45, delay: 0.25 + index * 0.08 }}
          animate={{ y: [0, index % 2 === 0 ? -4 : 4, 0] }}
        >
          {node.label}
        </motion.div>
      ))}
      <div className="absolute bottom-5 left-5 right-5 grid gap-3 md:grid-cols-3">
        {[
          ["30 days", "MVP direction"],
          ["2x", "faster lead response"],
          ["24/7", "AI enquiry capture"]
        ].map(([value, label], index) => (
          <motion.div
            key={label}
            initial={{ opacity: 0, y: 18 }}
            whileInView={{ opacity: 1, y: 0 }}
            viewport={{ once: true }}
            transition={{ duration: 0.45, delay: 0.45 + index * 0.08 }}
            className="border border-border bg-background/90 p-4 backdrop-blur"
          >
            <p className="text-2xl font-semibold text-foreground">{value}</p>
            <p className="text-xs uppercase tracking-[0.18em] text-muted-foreground">{label}</p>
          </motion.div>
        ))}
      </div>
    </motion.div>
  );
}

export function FeatureGrid({ items }: { items: FeatureItem[] }) {
  const icons = [BrainCircuit, Workflow, Cpu, Database, Network, LockKeyhole];

  return (
    <motion.div
      initial="hidden"
      whileInView="visible"
      viewport={{ once: true, margin: "-80px" }}
      variants={stagger}
      className="grid gap-4 md:grid-cols-2 lg:grid-cols-3"
    >
      {items.map((item, index) => {
        const Icon = icons[index % icons.length];
        return (
          <motion.article
            key={item.title}
            variants={fadeUp}
            transition={{ duration: 0.48 }}
            whileHover={{ y: -6, scale: 1.01 }}
            className="border border-border bg-background p-6 shadow-elegant"
          >
            <div className="mb-6 flex h-11 w-11 items-center justify-center border border-primary/20 bg-primary/10 text-primary">
              <Icon className="h-5 w-5" />
            </div>
            {item.meta ? <p className="mb-2 text-xs font-semibold uppercase tracking-[0.2em] text-accent-hover">{item.meta}</p> : null}
            <h3 className="text-xl font-semibold text-foreground">{item.title}</h3>
            <p className="mt-3 leading-7 text-muted-foreground">{item.description}</p>
          </motion.article>
        );
      })}
    </motion.div>
  );
}

export function StatStrip({ stats }: { stats: StatItem[] }) {
  return (
    <motion.div
      initial="hidden"
      whileInView="visible"
      viewport={{ once: true, margin: "-80px" }}
      variants={stagger}
      className="grid border border-border bg-background text-foreground shadow-elegant md:grid-cols-4"
    >
      {stats.map((stat) => (
        <motion.div key={stat.label} variants={fadeUp} transition={{ duration: 0.45 }} className="border-border p-6 md:border-r md:last:border-r-0">
          <p className="text-3xl font-semibold">{stat.value}</p>
          <p className="mt-2 text-sm text-muted-foreground">{stat.label}</p>
        </motion.div>
      ))}
    </motion.div>
  );
}

export function ImageFeature({
  image,
  eyebrow,
  title,
  description,
  points,
  flip = false
}: {
  image: string;
  eyebrow: string;
  title: string;
  description: string;
  points: string[];
  flip?: boolean;
}) {
  return (
    <div className="grid gap-8 lg:grid-cols-2 lg:items-center">
      <motion.div
        initial={{ opacity: 0, x: flip ? 36 : -36 }}
        whileInView={{ opacity: 1, x: 0 }}
        viewport={{ once: true, margin: "-80px" }}
        transition={{ duration: 0.65 }}
        className={cn("relative min-h-[360px] overflow-hidden border border-border", flip && "lg:order-2")}
      >
        <Image src={image} alt={title} fill className="object-cover" sizes="(min-width: 1024px) 50vw, 100vw" />
        <div className="absolute inset-0 bg-[linear-gradient(90deg,hsl(var(--foreground)/0.72),transparent_72%)]" />
        <div className="absolute bottom-6 left-6 max-w-xs border border-background/20 bg-background/90 p-4 text-foreground backdrop-blur">
          <p className="text-xs font-semibold uppercase tracking-[0.2em] text-primary">Business outcome</p>
          <p className="mt-2 text-sm leading-6 text-muted-foreground">Turn attention into enquiries, ideas into products, and workflows into automation.</p>
        </div>
      </motion.div>
      <motion.div
        initial="hidden"
        whileInView="visible"
        viewport={{ once: true, margin: "-80px" }}
        variants={stagger}
        className="space-y-5"
      >
        <Eyebrow>{eyebrow}</Eyebrow>
        <motion.h2 variants={fadeUp} transition={{ duration: 0.55 }} className="text-balance text-3xl font-semibold tracking-tight md:text-5xl">
          {title}
        </motion.h2>
        <motion.p variants={fadeUp} transition={{ duration: 0.55 }} className="text-lg leading-8 text-muted-foreground">
          {description}
        </motion.p>
        <div className="grid gap-3">
          {points.map((point) => (
            <motion.div key={point} variants={fadeUp} transition={{ duration: 0.45 }} className="flex gap-3 border border-border bg-background p-4">
              <GitBranch className="mt-1 h-4 w-4 shrink-0 text-primary" />
              <span className="text-sm leading-6 text-foreground">{point}</span>
            </motion.div>
          ))}
        </div>
      </motion.div>
    </div>
  );
}

export function VideoFeature({
  video,
  eyebrow,
  title,
  description,
  points,
  flip = false,
  outcome = "Turn attention into enquiries, ideas into products, and workflows into automation."
}: {
  video: string;
  eyebrow: string;
  title: string;
  description: string;
  points: string[];
  flip?: boolean;
  outcome?: string;
}) {
  return (
    <div className="grid gap-8 lg:grid-cols-2 lg:items-center">
      <motion.div
        initial={{ opacity: 0, x: flip ? 36 : -36 }}
        whileInView={{ opacity: 1, x: 0 }}
        viewport={{ once: true, margin: "-80px" }}
        transition={{ duration: 0.65 }}
        className={cn("relative min-h-[380px] overflow-hidden border border-border bg-foreground", flip && "lg:order-2")}
      >
        <video
          className="absolute inset-0 h-full w-full object-cover"
          src={video}
          autoPlay
          muted
          loop
          playsInline
          preload="metadata"
          aria-label={title}
        />
        <div className="absolute inset-0 bg-[linear-gradient(90deg,hsl(var(--foreground)/0.42),hsl(var(--foreground)/0.08)_58%,transparent)]" />
        <div className="absolute inset-0 site-grid opacity-[0.14] mix-blend-screen" />
        <motion.div
          animate={{ x: ["-18%", "118%"] }}
          transition={{ duration: 5.5, repeat: Infinity, ease: "easeInOut" }}
          className="absolute top-0 h-full w-1/3 bg-[linear-gradient(90deg,transparent,hsl(var(--background)/0.1),transparent)]"
        />
        <motion.div
          animate={{ y: [0, -8, 0], opacity: [0.82, 1, 0.82] }}
          transition={{ duration: 3.8, repeat: Infinity, ease: "easeInOut" }}
          className="absolute bottom-6 left-6 max-w-xs border border-background/20 bg-background/80 p-4 text-foreground backdrop-blur"
        >
          <p className="text-xs font-semibold uppercase tracking-[0.2em] text-primary">Business outcome</p>
          <p className="mt-2 text-sm leading-6 text-muted-foreground">{outcome}</p>
        </motion.div>
      </motion.div>
      <motion.div
        initial="hidden"
        whileInView="visible"
        viewport={{ once: true, margin: "-80px" }}
        variants={stagger}
        className="space-y-5"
      >
        <Eyebrow>{eyebrow}</Eyebrow>
        <motion.h2 variants={fadeUp} transition={{ duration: 0.55 }} className="text-balance text-3xl font-semibold tracking-tight md:text-5xl">
          {title}
        </motion.h2>
        <motion.p variants={fadeUp} transition={{ duration: 0.55 }} className="text-lg leading-8 text-muted-foreground">
          {description}
        </motion.p>
        <div className="grid gap-3">
          {points.map((point) => (
            <motion.div key={point} variants={fadeUp} transition={{ duration: 0.45 }} className="flex gap-3 border border-border bg-background p-4">
              <GitBranch className="mt-1 h-4 w-4 shrink-0 text-primary" />
              <span className="text-sm leading-6 text-foreground">{point}</span>
            </motion.div>
          ))}
        </div>
      </motion.div>
    </div>
  );
}

export function CtaBand({
  title = "Ready to turn an AI idea into business?",
  description = "Share the product, website, automation, or software service goal you have in mind. Lumentix will help you shape the next step clearly."
}: {
  title?: string;
  description?: string;
}) {
  return (
    <section className="border-y border-border gradient-hero py-16 text-foreground">
      <motion.div
        initial="hidden"
        whileInView="visible"
        viewport={{ once: true, margin: "-80px" }}
        variants={stagger}
        className="container mx-auto grid gap-8 px-4 lg:grid-cols-[1fr_auto] lg:items-center lg:px-8"
      >
        <motion.div variants={fadeUp} transition={{ duration: 0.55 }} className="max-w-3xl">
          <p className="text-sm font-semibold uppercase tracking-[0.22em] text-primary">Start the conversation</p>
          <h2 className="mt-4 text-balance text-3xl font-semibold md:text-5xl">{title}</h2>
          <p className="mt-4 text-lg leading-8 text-muted-foreground">{description}</p>
        </motion.div>
        <motion.div variants={fadeUp} transition={{ duration: 0.55 }} className="flex flex-col gap-3 sm:flex-row lg:flex-col">
          <Link
            href="/contact"
            className="inline-flex h-11 items-center justify-center border border-accent bg-accent px-5 text-sm font-semibold text-accent-foreground transition-colors hover:bg-accent-hover"
          >
            Start a project <ArrowRight className="ml-2 h-4 w-4" />
          </Link>
          <Link
            href="/login"
            className="inline-flex h-11 items-center justify-center border border-border bg-background px-5 text-sm font-semibold text-foreground transition-colors hover:bg-muted"
          >
            Employee Login
          </Link>
        </motion.div>
      </motion.div>
    </section>
  );
}
