// lumentix-complete/frontend/components/site/EnquiryForm.tsx
"use client";

import { type ChangeEvent, type FormEvent, useState } from "react";
import emailjs from "@emailjs/browser";
import { motion } from "framer-motion";
import { Mail, MessageSquare, Phone, User } from "lucide-react";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Textarea } from "@/components/ui/textarea";
import { useToast } from "@/hooks/use-toast";

const EnquiryForm = () => {
  const { toast } = useToast();
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [formData, setFormData] = useState({
    name: "",
    email: "",
    phone: "",
    aiIdea: "",
    message: ""
  });

  const handleChange = (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setFormData((current) => ({ ...current, [event.target.name]: event.target.value }));
  };

  const handleSubmit = async (event: FormEvent) => {
    event.preventDefault();
    setIsSubmitting(true);

    try {
      await emailjs.send(
        "service_5sbjkmn",
        "template_bp3qw19",
        {
          name: formData.name,
          email: formData.email,
          phone: formData.phone,
          ai_idea: formData.aiIdea,
          message: formData.message
        },
        "2AwcBSJm5czOHroBG"
      );

      toast({
        title: "Enquiry sent",
        description: "We will get back to you within 24 hours."
      });

      setFormData({ name: "", email: "", phone: "", aiIdea: "", message: "" });
    } catch (error) {
      console.error("EmailJS Error:", error);
      toast({
        title: "Could not send enquiry",
        description: "Please try again later or email our team directly.",
        variant: "destructive"
      });
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <section className="page-band bg-surface py-20">
      <div className="container mx-auto grid gap-10 px-4 lg:grid-cols-[0.9fr_1.1fr] lg:items-start lg:px-8">
        <motion.div
          initial={{ opacity: 0, x: -28 }}
          whileInView={{ opacity: 1, x: 0 }}
          viewport={{ once: true, margin: "-80px" }}
          transition={{ duration: 0.55 }}
        >
          <p className="text-sm font-semibold uppercase tracking-[0.22em] text-primary">Contact Lumentix</p>
          <h2 className="mt-5 text-balance text-3xl font-semibold md:text-5xl">
            Tell us your AI product idea, software need, or business challenge.
          </h2>
          <p className="mt-5 text-lg leading-8 text-muted-foreground">
            Share the goal, timeline, and current challenge. We will respond with a practical next step for your website, product, software service, or AI workflow.
          </p>
          <div className="mt-8 grid gap-4 text-sm text-muted-foreground">
            <p className="flex gap-3">
              <Mail className="mt-0.5 h-4 w-4 shrink-0 text-primary" />
              info@lumentix.in
            </p>
            <p className="flex gap-3">
              <Phone className="mt-0.5 h-4 w-4 shrink-0 text-primary" />
              +91 94945 48249
            </p>
          </div>
        </motion.div>

        <motion.form
          initial={{ opacity: 0, y: 28 }}
          whileInView={{ opacity: 1, y: 0 }}
          viewport={{ once: true, margin: "-80px" }}
          transition={{ duration: 0.55 }}
          onSubmit={handleSubmit}
          className="border border-border bg-background p-6 shadow-elegant"
        >
          <div className="grid gap-5">
            <label className="grid gap-2 text-sm font-semibold text-foreground" htmlFor="name">
              <span className="flex items-center gap-2">
                <User className="h-4 w-4 text-primary" />
                Full Name
              </span>
              <Input id="name" name="name" value={formData.name} onChange={handleChange} placeholder="Your name" required />
            </label>

            <label className="grid gap-2 text-sm font-semibold text-foreground" htmlFor="email">
              <span className="flex items-center gap-2">
                <Mail className="h-4 w-4 text-primary" />
                Email Address
              </span>
              <Input id="email" name="email" type="email" value={formData.email} onChange={handleChange} placeholder="you@example.com" required />
            </label>

            <label className="grid gap-2 text-sm font-semibold text-foreground" htmlFor="phone">
              <span className="flex items-center gap-2">
                <Phone className="h-4 w-4 text-primary" />
                Phone Number
              </span>
              <Input id="phone" name="phone" type="tel" value={formData.phone} onChange={handleChange} placeholder="+91 XXXXX XXXXX" required />
            </label>

            <label className="grid gap-2 text-sm font-semibold text-foreground" htmlFor="aiIdea">
              <span className="flex items-center gap-2">
                <MessageSquare className="h-4 w-4 text-primary" />
                AI Products & Ideas
              </span>
              <Textarea
                id="aiIdea"
                name="aiIdea"
                value={formData.aiIdea}
                onChange={handleChange}
                placeholder="Describe the AI product, automation idea, smart website, app concept, or workflow you want to create."
                rows={5}
                required
              />
            </label>

            <label className="grid gap-2 text-sm font-semibold text-foreground" htmlFor="message">
              <span className="flex items-center gap-2">
                <MessageSquare className="h-4 w-4 text-primary" />
                Additional Message
              </span>
              <Textarea
                id="message"
                name="message"
                value={formData.message}
                onChange={handleChange}
                placeholder="Add timeline, budget range, preferred service, or anything else we should know."
                rows={4}
              />
            </label>

            <Button type="submit" size="lg" disabled={isSubmitting} className="w-full bg-primary text-primary-foreground hover:bg-primary-hover">
              {isSubmitting ? "Sending..." : "Send Enquiry"}
            </Button>
          </div>
        </motion.form>
      </div>
    </section>
  );
};

export default EnquiryForm;
