// lumentix-complete/frontend/components/site/Team.tsx
"use client";

import Image from "next/image";
import { motion, type Variants } from "framer-motion";

interface TeamMember {
  name: string;
  position: string;
  photo: string;
}

const teamMembers: TeamMember[] = [
  {
    name: "Manoj Kumar",
    position: "CEO",
    photo: "/tem.jpg",
  },
  {
    name: "Sandeep",
    position: "CFO",
    photo: "/careergrowth.png",
  },
  {
    name: "Random",
    position: "CTO",
    photo: "/flexiblework.png",
  },
];

const Team = () => {
  const cardVariants: Variants = {
    hidden: { opacity: 0, x: 50 },
    visible: (i: number) => ({
      opacity: 1,
      x: 0,
      transition: {
        delay: i * 0.3, // Staggered animation
        type: "spring",
        stiffness: 100,
      },
    }),
  };

  return (
    <section className="py-16 bg-gray-50">
      <div className="container mx-auto px-4 lg:px-8 text-center">
        <h2 className="text-4xl font-bold mb-8">Our Leadership Team</h2>
        <p className="text-gray-600 mb-12 max-w-2xl mx-auto">
          Meet the visionary leaders who drive our mission and shape the future
          of Lumentix Solutions.
        </p>

        <div className="flex justify-center">
          <div className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 gap-8">
            {teamMembers.map((member, index) => (
              <motion.div
                key={index}
                custom={index}
                initial="hidden"
                whileInView="visible"
                viewport={{ once: true, amount: 0.5 }}
                variants={cardVariants}
                className="mx-auto max-w-xs overflow-hidden border border-border bg-white shadow-lg transition-shadow duration-300 hover:shadow-2xl"
              >
                <div className="relative w-full h-64">
                  <Image
                    src={member.photo}
                    alt={member.name}
                    fill
                    className="object-cover"
                    sizes="(min-width: 1024px) 320px, 100vw"
                  />
                </div>
                <div className="p-6 text-center">
                  <h3 className="text-xl font-semibold">{member.name}</h3>
                  <p className="text-gray-500">{member.position}</p>
                </div>
              </motion.div>
            ))}
          </div>
        </div>
      </div>
    </section>
  );
};

export default Team;
