// lumentix-complete/frontend/next.config.mjs
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const configuredBasePath = process.env.NEXT_PUBLIC_BASE_PATH ?? "";
const normalizedBasePath =
  configuredBasePath && configuredBasePath !== "/"
    ? `/${configuredBasePath.replace(/^\/+|\/+$/g, "")}`
    : "";

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  ...(normalizedBasePath ? { basePath: normalizedBasePath } : {}),
  async redirects() {
    if (!normalizedBasePath) return [];

    return [
      {
        source: "/",
        destination: normalizedBasePath,
        permanent: false,
        basePath: false
      },
      {
        source: "/verify/:path*",
        destination: `${normalizedBasePath}/verify/:path*`,
        permanent: false,
        basePath: false
      },
      {
        source: "/dashboard/:path*",
        destination: `${normalizedBasePath}/dashboard/:path*`,
        permanent: false,
        basePath: false
      },
      {
        source: "/interns/:path*",
        destination: `${normalizedBasePath}/interns/:path*`,
        permanent: false,
        basePath: false
      },
      {
        source: "/documents/:path*",
        destination: `${normalizedBasePath}/documents/:path*`,
        permanent: false,
        basePath: false
      },
      {
        source: "/applicants/:path*",
        destination: `${normalizedBasePath}/applicants/:path*`,
        permanent: false,
        basePath: false
      },
      {
        source: "/bgv/:path*",
        destination: `${normalizedBasePath}/bgv/:path*`,
        permanent: false,
        basePath: false
      },
      {
        source: "/system/:path*",
        destination: `${normalizedBasePath}/system/:path*`,
        permanent: false,
        basePath: false
      },
      {
        source: "/login",
        destination: `${normalizedBasePath}/login`,
        permanent: false,
        basePath: false
      }
    ];
  },
  turbopack: {
    root: __dirname
  }
};

export default nextConfig;
