// lumentix-complete/frontend/app/(protected)/documents/generate/page.tsx
"use client";

import { useCallback, useEffect, useMemo, useState } from "react";
import { Download } from "lucide-react";
import { Button } from "@/components/ui/button";
import { Select } from "@/components/ui/select";
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from "@/components/ui/table";
import { apiFetch, downloadAuthenticatedFile } from "@/lib/api";
import type { DocumentType, Intern, InternDocument } from "@/lib/types";

const documentTypes: DocumentType[] = ["offer", "certificate", "experience", "relieving", "nda"];
const documentLabels: Record<DocumentType, string> = {
  offer: "Offer Letter",
  certificate: "Completion Certificate",
  experience: "Experience Letter",
  relieving: "Relieving Letter",
  nda: "NDA"
};

type DocumentWithIntern = InternDocument & {
  intern: {
    id: string;
    name: string;
    reference_id: string;
  };
};

type DocumentMeta = {
  total: number;
  offset: number;
  limit: number;
};

function getDocumentWindow(page: number) {
  if (page <= 1) {
    return { offset: 0, limit: 10 };
  }

  return {
    offset: 10 + (page - 2) * 20,
    limit: 20
  };
}

function getDocumentPages(total: number) {
  if (total <= 10) return 1;
  return 1 + Math.ceil((total - 10) / 20);
}

export default function GenerateDocumentsPage() {
  const [interns, setInterns] = useState<Intern[]>([]);
  const [documents, setDocuments] = useState<DocumentWithIntern[]>([]);
  const [documentMeta, setDocumentMeta] = useState<DocumentMeta>({ total: 0, offset: 0, limit: 10 });
  const [documentPage, setDocumentPage] = useState(1);
  const [internId, setInternId] = useState("");
  const [type, setType] = useState<DocumentType>("offer");
  const [busy, setBusy] = useState(false);
  const [documentsLoading, setDocumentsLoading] = useState(true);
  const [message, setMessage] = useState("");
  const totalDocumentPages = useMemo(() => getDocumentPages(documentMeta.total), [documentMeta.total]);

  const load = useCallback(async (targetPage = documentPage) => {
    const window = getDocumentWindow(targetPage);
    setDocumentsLoading(true);

    try {
      const [internResponse, documentResponse] = await Promise.all([
        apiFetch<{ data: Intern[] }>("/api/interns?limit=100"),
        apiFetch<{ data: DocumentWithIntern[]; meta: DocumentMeta }>(
          `/api/documents?offset=${window.offset}&limit=${window.limit}`
        )
      ]);
      setInterns(internResponse.data);
      setDocuments(documentResponse.data);
      setDocumentMeta(documentResponse.meta);
      setInternId((current) => current || internResponse.data[0]?.id || "");
    } catch (err) {
      setMessage(err instanceof Error ? err.message : "Unable to load generated documents");
    } finally {
      setDocumentsLoading(false);
    }
  }, [documentPage]);

  useEffect(() => {
    load(documentPage);
  }, [documentPage, load]);

  async function generate(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    setBusy(true);
    setMessage("");

    try {
      await apiFetch("/api/documents/generate", {
        method: "POST",
        body: JSON.stringify({ internId, type })
      });
      setMessage("Document saved. The latest 10 generated files are shown first.");
      setDocumentPage(1);
      await load(1);
    } catch (err) {
      setMessage(err instanceof Error ? err.message : "Unable to generate document");
    } finally {
      setBusy(false);
    }
  }

  async function downloadDocument(document: DocumentWithIntern) {
    const filename = `${document.intern.reference_id}-${document.type}.pdf`;
    await downloadAuthenticatedFile(`/api/documents/${document.id}/download`, filename);
  }

  return (
    <div className="space-y-6">
      <div>
        <p className="text-sm uppercase text-muted-foreground">Document center</p>
        <h1 className="text-3xl font-semibold">Generate Documents</h1>
      </div>

      <form onSubmit={generate} className="grid gap-3 rounded-lg border border-border bg-background p-5 shadow-soft md:grid-cols-[1fr_220px_auto]">
        <Select value={internId} onChange={(event) => setInternId(event.target.value)} required>
          {interns.length === 0 ? <option value="">No records available</option> : null}
          {interns.map((intern) => (
            <option key={intern.id} value={intern.id}>
              {intern.name} - {intern.reference_id}
            </option>
          ))}
        </Select>
        <Select value={type} onChange={(event) => setType(event.target.value as DocumentType)}>
          {documentTypes.map((item) => (
            <option key={item} value={item}>
              {documentLabels[item]}
            </option>
          ))}
        </Select>
        <Button disabled={busy || !internId}>{busy ? "Generating..." : "Generate PDF"}</Button>
        {message ? <p className="md:col-span-3 text-sm text-muted-foreground">{message}</p> : null}
      </form>

      <section className="rounded-lg border border-border bg-background shadow-soft">
        <div className="border-b border-border p-5">
          <h2 className="text-base font-semibold">Generated Files</h2>
          <p className="mt-1 text-sm text-muted-foreground">
            Page 1 shows the latest 10 generated files. Older files are shown 20 per page.
          </p>
        </div>
        <div className="flex flex-col gap-3 border-b border-border p-5 md:flex-row md:items-center md:justify-between">
          <p className="text-sm text-muted-foreground">
            Showing {documents.length === 0 ? 0 : documentMeta.offset + 1}
            {"-"}
            {Math.min(documentMeta.offset + documents.length, documentMeta.total)} of {documentMeta.total} generated files.
          </p>
          <div className="flex gap-2">
            <Button type="button" variant="outline" disabled={documentsLoading || documentPage <= 1} onClick={() => setDocumentPage((current) => Math.max(1, current - 1))}>
              Previous
            </Button>
            <Button type="button" variant="outline" disabled={documentsLoading || documentPage >= totalDocumentPages} onClick={() => setDocumentPage((current) => current + 1)}>
              Next
            </Button>
          </div>
        </div>
        <Table>
          <TableHeader>
            <TableRow>
              <TableHead>Person</TableHead>
              <TableHead>Reference</TableHead>
              <TableHead>Type</TableHead>
              <TableHead>Created</TableHead>
              <TableHead className="text-right">Download</TableHead>
            </TableRow>
          </TableHeader>
          <TableBody>
            {documentsLoading ? (
              <TableRow>
                <TableCell colSpan={5} className="text-center text-muted-foreground">
                  Loading generated documents...
                </TableCell>
              </TableRow>
            ) : documents.length === 0 ? (
              <TableRow>
                <TableCell colSpan={5} className="text-center text-muted-foreground">
                  No generated documents.
                </TableCell>
              </TableRow>
            ) : (
              documents.map((document) => (
                <TableRow key={document.id}>
                  <TableCell>{document.intern.name}</TableCell>
                  <TableCell className="font-mono text-xs">{document.intern.reference_id}</TableCell>
                  <TableCell>{documentLabels[document.type]}</TableCell>
                  <TableCell>{new Date(document.created_at).toLocaleString()}</TableCell>
                  <TableCell className="text-right">
                    <button
                      type="button"
                      className="inline-flex h-8 items-center justify-center rounded-lg border border-border px-3 text-xs font-medium hover:bg-muted"
                      onClick={() => downloadDocument(document)}
                    >
                      <Download className="mr-2 h-3 w-3" />
                      PDF
                    </button>
                  </TableCell>
                </TableRow>
              ))
            )}
          </TableBody>
        </Table>
      </section>
    </div>
  );
}
