// lumentix-complete/frontend/components/interns/intern-form.tsx
"use client";

import { useState } from "react";
import { useRouter } from "next/navigation";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { Select } from "@/components/ui/select";
import { apiFetch } from "@/lib/api";
import { calculateDurationLabel } from "@/lib/duration";
import type { Intern, InternStatus } from "@/lib/types";

const statuses: InternStatus[] = ["Applied", "Offered", "Joined", "Completed", "Relieved"];

type FormState = {
  name: string;
  email: string;
  phone: string;
  role: string;
  department: string;
  joining_date: string;
  exit_date: string;
  duration: string;
  manager: string;
  status: InternStatus;
};

const emptyState: FormState = {
  name: "",
  email: "",
  phone: "",
  role: "",
  department: "",
  joining_date: "",
  exit_date: "",
  duration: "",
  manager: "",
  status: "Applied"
};

export function InternForm({ initial }: { initial?: Intern }) {
  const router = useRouter();
  const [form, setForm] = useState<FormState>(
    initial
      ? {
          name: initial.name,
          email: initial.email,
          phone: initial.phone,
          role: initial.role,
          department: initial.department,
          joining_date: initial.joining_date.slice(0, 10),
          exit_date: initial.exit_date ? initial.exit_date.slice(0, 10) : "",
          duration: initial.duration,
          manager: initial.manager,
          status: initial.status
        }
      : emptyState
  );
  const [saving, setSaving] = useState(false);
  const [error, setError] = useState("");
  const invalidDuration = form.duration.startsWith("Exit date cannot");

  function update<K extends keyof FormState>(key: K, value: FormState[K]) {
    setForm((current) => {
      const normalizedValue = key === "phone" ? String(value).replace(/\D/g, "") : value;
      const next = { ...current, [key]: normalizedValue };
      if (key === "joining_date" || key === "exit_date") {
        next.duration = calculateDurationLabel(next.joining_date, next.exit_date);
      }
      return next;
    });
  }

  async function submit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    setSaving(true);
    setError("");

    try {
      if (initial) {
        await apiFetch(`/api/intern/${initial.id}`, {
          method: "PUT",
          body: JSON.stringify(form)
        });
        router.refresh();
      } else {
        const response = await apiFetch<{ data: Intern }>("/api/intern", {
          method: "POST",
          body: JSON.stringify(form)
        });
        router.push(`/interns/${response.data.id}`);
      }
    } catch (err) {
      setError(err instanceof Error ? err.message : "Unable to save record");
    } finally {
      setSaving(false);
    }
  }

  return (
    <form onSubmit={submit} className="rounded-lg border border-border bg-background p-5 shadow-soft">
      <div className="grid gap-4 md:grid-cols-2">
        <Field label="Name">
          <Input value={form.name} onChange={(event) => update("name", event.target.value)} pattern="[A-Za-z .'-]{2,120}" title="Use letters only." required />
        </Field>
        <Field label="Email">
          <Input type="email" value={form.email} onChange={(event) => update("email", event.target.value)} required />
        </Field>
        <Field label="Phone">
          <Input value={form.phone} onChange={(event) => update("phone", event.target.value)} inputMode="numeric" pattern="[0-9]{7,15}" title="Use digits only, 7 to 15 numbers." required />
        </Field>
        <Field label="Role">
          <Input value={form.role} onChange={(event) => update("role", event.target.value)} minLength={2} maxLength={120} required />
        </Field>
        <Field label="Department">
          <Input value={form.department} onChange={(event) => update("department", event.target.value)} minLength={2} maxLength={120} required />
        </Field>
        <Field label="Joining Date">
          <Input type="date" value={form.joining_date} onChange={(event) => update("joining_date", event.target.value)} required />
        </Field>
        <Field label="Date Of Exit">
          <Input type="date" value={form.exit_date} onChange={(event) => update("exit_date", event.target.value)} />
        </Field>
        <Field label="Duration">
          <Input value={form.duration} placeholder="Auto calculated" readOnly required />
        </Field>
        <Field label="Manager">
          <Input value={form.manager} onChange={(event) => update("manager", event.target.value)} pattern="[A-Za-z .'-]{2,120}" title="Use letters only." required />
        </Field>
        <Field label="Status">
          <Select value={form.status} onChange={(event) => update("status", event.target.value as InternStatus)}>
            {statuses.map((status) => (
              <option key={status} value={status}>
                {status}
              </option>
            ))}
          </Select>
        </Field>
      </div>
      {invalidDuration ? <p className="mt-4 text-sm text-destructive">{form.duration}</p> : null}
      {error ? <p className="mt-4 text-sm text-destructive">{error}</p> : null}
      <div className="mt-6 flex justify-end gap-2">
        <Button type="button" variant="outline" onClick={() => router.back()}>
          Cancel
        </Button>
        <Button type="submit" disabled={saving || invalidDuration}>
          {saving ? "Saving..." : initial ? "Save Changes" : "Add Record"}
        </Button>
      </div>
    </form>
  );
}

function Field({ label, children }: { label: string; children: React.ReactNode }) {
  return (
    <div className="space-y-2">
      <Label>{label}</Label>
      {children}
    </div>
  );
}
