// lumentix-complete/frontend/components/interns/intern-table.tsx
"use client";

import Link from "next/link";
import { useCallback, useEffect, useMemo, useState } from "react";
import { Upload } from "lucide-react";
import { Button } from "@/components/ui/button";
import { Dialog, DialogContent, DialogDescription, DialogHeader, DialogTitle, DialogTrigger } from "@/components/ui/dialog";
import { Input } from "@/components/ui/input";
import { Select } from "@/components/ui/select";
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from "@/components/ui/table";
import { StatusBadge } from "@/components/status-badge";
import { apiFetch } from "@/lib/api";
import type { Intern, InternStatus } from "@/lib/types";

const statuses: Array<InternStatus | "All"> = ["All", "Applied", "Offered", "Joined", "Completed", "Relieved"];

export function InternTable() {
  const [interns, setInterns] = useState<Intern[]>([]);
  const [search, setSearch] = useState("");
  const [status, setStatus] = useState<InternStatus | "All">("All");
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");

  const query = useMemo(() => {
    const params = new URLSearchParams({ page: "1", limit: "50" });
    if (search) params.set("search", search);
    if (status !== "All") params.set("status", status);
    return params.toString();
  }, [search, status]);

  const loadInterns = useCallback(async () => {
    setLoading(true);
    setError("");
    try {
      const response = await apiFetch<{ data: Intern[] }>(`/api/interns?${query}`);
      setInterns(response.data);
    } catch (err) {
      setError(err instanceof Error ? err.message : "Unable to load records");
    } finally {
      setLoading(false);
    }
  }, [query]);

  useEffect(() => {
    loadInterns();
  }, [loadInterns]);

  return (
    <div className="rounded-lg border border-border bg-background shadow-soft">
      <div className="flex flex-col gap-3 border-b border-border p-4 md:flex-row md:items-center md:justify-between">
        <div className="grid gap-3 md:grid-cols-[280px_180px]">
          <Input placeholder="Search name, email, role, reference..." value={search} onChange={(event) => setSearch(event.target.value)} />
          <Select value={status} onChange={(event) => setStatus(event.target.value as InternStatus | "All")}>
            {statuses.map((item) => (
              <option key={item} value={item}>
                {item === "All" ? "All statuses" : item}
              </option>
            ))}
          </Select>
        </div>
        <div className="flex gap-2">
          <BulkUploadDialog onUploaded={loadInterns} />
          <Link className="inline-flex h-10 items-center justify-center rounded-lg bg-primary px-4 text-sm font-medium text-primary-foreground hover:bg-primary/90" href="/interns/add">
            Add Record
          </Link>
        </div>
      </div>
      {error ? <p className="p-4 text-sm text-destructive">{error}</p> : null}
      <Table>
        <TableHeader>
          <TableRow>
            <TableHead>Name</TableHead>
            <TableHead>Reference</TableHead>
            <TableHead>Role</TableHead>
            <TableHead>Department</TableHead>
            <TableHead>Status</TableHead>
            <TableHead>Documents</TableHead>
            <TableHead className="text-right">Action</TableHead>
          </TableRow>
        </TableHeader>
        <TableBody>
          {loading ? (
            <TableRow>
              <TableCell colSpan={7} className="text-center text-muted-foreground">
                Loading records...
              </TableCell>
            </TableRow>
          ) : interns.length === 0 ? (
            <TableRow>
              <TableCell colSpan={7} className="text-center text-muted-foreground">
                No records found.
              </TableCell>
            </TableRow>
          ) : (
            interns.map((intern) => (
              <TableRow key={intern.id}>
                <TableCell>
                  <div className="font-medium">{intern.name}</div>
                  <div className="text-xs text-muted-foreground">{intern.email}</div>
                </TableCell>
                <TableCell className="font-mono text-xs">{intern.reference_id}</TableCell>
                <TableCell>{intern.role}</TableCell>
                <TableCell>{intern.department}</TableCell>
                <TableCell>
                  <StatusBadge status={intern.status} />
                </TableCell>
                <TableCell>{intern.documents.length}</TableCell>
                <TableCell className="text-right">
                  <Link className="text-sm font-medium text-primary hover:underline" href={`/interns/${intern.id}`}>
                    Open
                  </Link>
                </TableCell>
              </TableRow>
            ))
          )}
        </TableBody>
      </Table>
    </div>
  );
}

function BulkUploadDialog({ onUploaded }: { onUploaded: () => void }) {
  const [file, setFile] = useState<File | null>(null);
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState("");
  const [busy, setBusy] = useState(false);

  async function uploadCsv(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault();
    if (!file) return;

    setBusy(true);
    setMessage("");

    try {
      const data = new FormData();
      data.set("file", file);
      const response = await apiFetch<{ meta: { created: number } }>("/api/interns/bulk-upload", {
        method: "POST",
        body: data
      });
      setMessage(`${response.meta.created} records imported.`);
      onUploaded();
    } catch (err) {
      setMessage(err instanceof Error ? err.message : "Upload failed");
    } finally {
      setBusy(false);
    }
  }

  return (
    <Dialog open={open} onOpenChange={setOpen}>
      <DialogTrigger asChild>
        <Button type="button" variant="outline">
          <Upload className="mr-2 h-4 w-4" />
          CSV Upload
        </Button>
      </DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>CSV Bulk Upload</DialogTitle>
          <DialogDescription>
            Columns: name, email, phone, role, department, joining_date, exit_date, manager, status. Duration is calculated automatically.
          </DialogDescription>
        </DialogHeader>
        <form className="space-y-4" onSubmit={uploadCsv}>
          <Input type="file" accept=".csv,text/csv" onChange={(event) => setFile(event.target.files?.[0] ?? null)} />
          {message ? <p className="text-sm text-muted-foreground">{message}</p> : null}
          <div className="flex justify-end gap-2">
            <Button type="button" variant="outline" onClick={() => setOpen(false)}>
              Close
            </Button>
            <Button type="submit" disabled={!file || busy}>
              {busy ? "Uploading..." : "Upload"}
            </Button>
          </div>
        </form>
      </DialogContent>
    </Dialog>
  );
}
