// lumentix-complete/frontend/app/(protected)/interns/[id]/page.tsx
"use client";

import Link from "next/link";
import { useParams } from "next/navigation";
import { useCallback, useEffect, useState } from "react";
import { Download, Mail, QrCode } from "lucide-react";
import { InternForm } from "@/components/interns/intern-form";
import { StatusBadge } from "@/components/status-badge";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { Select } from "@/components/ui/select";
import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from "@/components/ui/table";
import { apiFetch, downloadAuthenticatedFile } from "@/lib/api";
import type { ActivityLog, DocumentType, Intern, InternDocument, InternStatus } from "@/lib/types";

const statuses: InternStatus[] = ["Applied", "Offered", "Joined", "Completed", "Relieved"];
const documentTypes: DocumentType[] = ["offer", "certificate", "experience", "relieving", "nda"];
const documentLabels: Record<DocumentType, string> = {
  offer: "Offer Letter",
  certificate: "Completion Certificate",
  experience: "Experience Letter",
  relieving: "Relieving Letter",
  nda: "NDA"
};

export default function InternDetailPage() {
  const params = useParams<{ id: string }>();
  const [intern, setIntern] = useState<Intern | null>(null);
  const [status, setStatus] = useState<InternStatus>("Applied");
  const [message, setMessage] = useState("");
  const [busy, setBusy] = useState(false);
  const [editing, setEditing] = useState(false);

  const loadIntern = useCallback(async () => {
    const response = await apiFetch<{ data: Intern }>(`/api/intern/${params.id}`);
    setIntern(response.data);
    setStatus(response.data.status);
  }, [params.id]);

  useEffect(() => {
    loadIntern();
  }, [loadIntern]);

  async function updateStatus() {
    if (!intern) return;
    setBusy(true);
    setMessage("");
    try {
      await apiFetch(`/api/intern/${intern.id}`, {
        method: "PUT",
        body: JSON.stringify({ status })
      });
      setMessage("Status updated.");
      await loadIntern();
    } catch (err) {
      setMessage(err instanceof Error ? err.message : "Unable to update status");
    } finally {
      setBusy(false);
    }
  }

  async function setRevoked(is_revoked: boolean) {
    if (!intern) return;
    setBusy(true);
    setMessage("");
    try {
      await apiFetch(`/api/intern/${intern.id}`, {
        method: "PUT",
        body: JSON.stringify({ is_revoked })
      });
      setMessage(is_revoked ? "Verification revoked." : "Verification restored.");
      await loadIntern();
    } catch (err) {
      setMessage(err instanceof Error ? err.message : "Unable to update verification");
    } finally {
      setBusy(false);
    }
  }

  async function generate(type: DocumentType) {
    if (!intern) return;
    setBusy(true);
    setMessage("");
    try {
      await apiFetch("/api/documents/generate", {
        method: "POST",
        body: JSON.stringify({ internId: intern.id, type })
      });
      setMessage(`${documentLabels[type]} saved. Downloads use the latest document data.`);
      await loadIntern();
    } catch (err) {
      setMessage(err instanceof Error ? err.message : "Unable to generate document");
    } finally {
      setBusy(false);
    }
  }

  async function emailOffer(documentId: string) {
    setBusy(true);
    setMessage("");
    try {
      await apiFetch(`/api/documents/${documentId}/email`, {
        method: "POST",
        body: JSON.stringify({})
      });
      setMessage("Offer letter email queued.");
    } catch (err) {
      setMessage(err instanceof Error ? err.message : "Unable to send email");
    } finally {
      setBusy(false);
    }
  }

  async function downloadDocument(document: InternDocument) {
    if (!intern) return;
    await downloadAuthenticatedFile(`/api/documents/${document.id}/download`, `${intern.reference_id}-${document.type}.pdf`);
  }

  if (!intern) {
    return <div className="text-sm text-muted-foreground">Loading record...</div>;
  }

  const bgvAccessRecords = getBgvAccessRecords(intern.activities);

  return (
    <div className="space-y-6">
      <div className="flex flex-col gap-3 md:flex-row md:items-end md:justify-between">
        <div>
          <p className="text-sm uppercase text-muted-foreground">Employment profile</p>
          <h1 className="text-3xl font-semibold">{intern.name}</h1>
          <p className="font-mono text-sm text-muted-foreground">{intern.reference_id}</p>
        </div>
        <Link
          className="inline-flex h-10 items-center justify-center rounded-lg border border-border bg-background px-4 text-sm font-medium hover:bg-muted"
          href={`/verify/${intern.reference_id}`}
        >
          <QrCode className="mr-2 h-4 w-4" />
          Public Verify
        </Link>
      </div>

      <div className="grid gap-4 lg:grid-cols-[1fr_360px]">
        <section className="rounded-lg border border-border bg-background p-5 shadow-soft">
          <div className="mb-5 flex items-center justify-between">
            <h2 className="text-base font-semibold">Lifecycle</h2>
            <StatusBadge status={intern.status} />
          </div>
          <div className="grid gap-3 md:grid-cols-2">
            <Info label="Email" value={intern.email} />
            <Info label="Phone" value={intern.phone} />
            <Info label="Role" value={intern.role} />
            <Info label="Department" value={intern.department} />
            <Info label="Joining Date" value={new Date(intern.joining_date).toLocaleDateString()} />
            <Info label="Date Of Exit" value={intern.exit_date ? new Date(intern.exit_date).toLocaleDateString() : "Ongoing"} />
            <Info label="Duration" value={intern.duration} />
            <Info label="Manager" value={intern.manager} />
            <div>
              <p className="text-xs uppercase text-muted-foreground">BGV Status</p>
              <Badge variant={intern.is_revoked ? "red" : "teal"}>{intern.is_revoked ? "Revoked" : "Valid"}</Badge>
            </div>
          </div>
          <div className="mt-5 grid gap-3 md:grid-cols-[1fr_auto_auto]">
            <Select value={status} onChange={(event) => setStatus(event.target.value as InternStatus)}>
              {statuses.map((item) => (
                <option key={item} value={item}>
                  {item}
                </option>
              ))}
            </Select>
            <Button onClick={updateStatus} disabled={busy}>
              Update Status
            </Button>
            <Button variant={intern.is_revoked ? "outline" : "destructive"} onClick={() => setRevoked(!intern.is_revoked)} disabled={busy}>
              {intern.is_revoked ? "Restore BGV" : "Revoke BGV"}
            </Button>
          </div>
          {message ? <p className="mt-3 text-sm text-muted-foreground">{message}</p> : null}
        </section>

        <section className="rounded-lg border border-border bg-background p-5 shadow-soft">
          <h2 className="mb-4 text-base font-semibold">Generate Documents</h2>
          <div className="grid gap-2">
            {documentTypes.map((type) => (
              <Button key={type} variant="outline" onClick={() => generate(type)} disabled={busy}>
                Generate {documentLabels[type]}
              </Button>
            ))}
          </div>
        </section>
      </div>

      <section className="rounded-lg border border-border bg-background shadow-soft">
        <div className="border-b border-border p-5">
          <h2 className="text-base font-semibold">Documents</h2>
        </div>
        <Table>
          <TableHeader>
            <TableRow>
              <TableHead>Type</TableHead>
              <TableHead>Created</TableHead>
              <TableHead className="text-right">Actions</TableHead>
            </TableRow>
          </TableHeader>
          <TableBody>
            {intern.documents.length === 0 ? (
              <TableRow>
                <TableCell colSpan={3} className="text-center text-muted-foreground">
                  No documents generated.
                </TableCell>
              </TableRow>
            ) : (
              intern.documents.map((document) => (
                <TableRow key={document.id}>
                  <TableCell>{documentLabels[document.type]}</TableCell>
                  <TableCell>{new Date(document.created_at).toLocaleString()}</TableCell>
                  <TableCell className="flex justify-end gap-2">
                    {document.type === "offer" ? (
                      <Button type="button" variant="outline" size="sm" onClick={() => emailOffer(document.id)} disabled={busy}>
                        <Mail className="mr-2 h-3 w-3" />
                        Email
                      </Button>
                    ) : null}
                    <button
                      type="button"
                      className="inline-flex h-8 items-center justify-center rounded-lg border border-border bg-background px-3 text-xs font-medium hover:bg-muted"
                      onClick={() => downloadDocument(document)}
                    >
                      <Download className="mr-2 h-3 w-3" />
                      Download
                    </button>
                  </TableCell>
                </TableRow>
              ))
            )}
          </TableBody>
        </Table>
      </section>

      <section className="rounded-lg border border-border bg-background shadow-soft">
        <div className="border-b border-border p-5">
          <h2 className="text-base font-semibold">BGV Access Records</h2>
          <p className="mt-1 text-sm text-muted-foreground">
            Every public verification form submission is saved here with verifier and request details.
          </p>
        </div>
        <Table>
          <TableHeader>
            <TableRow>
              <TableHead>Accessed At</TableHead>
              <TableHead>Verifier</TableHead>
              <TableHead>Organization</TableHead>
              <TableHead>Purpose</TableHead>
              <TableHead>Result</TableHead>
              <TableHead>Request Details</TableHead>
            </TableRow>
          </TableHeader>
          <TableBody>
            {bgvAccessRecords.length === 0 ? (
              <TableRow>
                <TableCell colSpan={6} className="text-center text-muted-foreground">
                  No public BGV verification access has been recorded yet.
                </TableCell>
              </TableRow>
            ) : (
              bgvAccessRecords.map((record) => (
                <TableRow key={record.id}>
                  <TableCell className="whitespace-nowrap text-sm">{new Date(record.createdAt).toLocaleString()}</TableCell>
                  <TableCell>
                    <div className="text-sm font-medium">{record.verifierName}</div>
                    <div className="text-xs text-muted-foreground">{record.email}</div>
                    {record.phone ? <div className="text-xs text-muted-foreground">{record.phone}</div> : null}
                  </TableCell>
                  <TableCell>
                    <div className="text-sm">{record.organization}</div>
                    {record.designation ? <div className="text-xs text-muted-foreground">{record.designation}</div> : null}
                  </TableCell>
                  <TableCell>
                    <div className="text-sm">{record.purpose}</div>
                    {record.actionLabel ? <div className="text-xs text-muted-foreground">{record.actionLabel}</div> : null}
                    {record.notes ? <div className="mt-1 text-xs text-muted-foreground">{record.notes}</div> : null}
                  </TableCell>
                  <TableCell>
                    <Badge variant={record.result === "Valid" ? "teal" : "red"}>{record.result}</Badge>
                  </TableCell>
                  <TableCell>
                    <div className="text-xs text-muted-foreground">IP: {record.ipAddress}</div>
                    <div className="max-w-[260px] truncate text-xs text-muted-foreground" title={record.userAgent}>
                      {record.userAgent}
                    </div>
                  </TableCell>
                </TableRow>
              ))
            )}
          </TableBody>
        </Table>
      </section>

      <section className="grid gap-4 lg:grid-cols-[1fr_360px]">
        <div className="rounded-lg border border-border bg-background p-5 shadow-soft">
          <div className="flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between">
            <div>
              <h2 className="text-base font-semibold">Document Data</h2>
              <p className="text-sm text-muted-foreground">Saved data is used whenever a PDF is downloaded or emailed.</p>
            </div>
            <Button type="button" variant={editing ? "outline" : "default"} onClick={() => setEditing((current) => !current)}>
              {editing ? "Close Edit" : "Edit Document Data"}
            </Button>
          </div>
          {editing ? (
            <div className="mt-5">
              <InternForm initial={intern} />
            </div>
          ) : null}
        </div>
        <div className="rounded-lg border border-border bg-background p-5 shadow-soft">
          <h2 className="mb-4 text-base font-semibold">Activity Timeline</h2>
          <div className="space-y-4">
            {(intern.activities ?? []).length === 0 ? (
              <p className="text-sm text-muted-foreground">No activity yet.</p>
            ) : (
              intern.activities?.map((activity) => (
                <div key={activity.id} className="border-l-2 border-primary pl-3">
                  <p className="text-sm font-medium">{activity.action.replaceAll("_", " ")}</p>
                  <p className="text-xs text-muted-foreground">{new Date(activity.created_at).toLocaleString()}</p>
                  <ActivityMetadata metadata={activity.metadata} />
                </div>
              ))
            )}
          </div>
        </div>
      </section>
    </div>
  );
}

type BgvAccessRecord = {
  id: string;
  actionLabel: string;
  verifierName: string;
  email: string;
  phone: string;
  organization: string;
  designation: string;
  purpose: string;
  notes: string;
  result: "Valid" | "Revoked";
  ipAddress: string;
  userAgent: string;
  createdAt: string;
};

function readString(value: unknown, fallback = "Unavailable") {
  return typeof value === "string" && value.trim().length > 0 ? value : fallback;
}

function getBgvAccessRecords(activities: ActivityLog[] = []): BgvAccessRecord[] {
  return activities
    .filter((activity) =>
      ["PUBLIC_VERIFICATION_COMPLETED", "PUBLIC_VERIFICATION_SUMMARY_DOWNLOADED"].includes(activity.action)
    )
    .map((activity) => {
      const metadata = activity.metadata ?? {};
      const verifier = metadata.verifier;
      const details = typeof verifier === "object" && verifier !== null ? verifier as Record<string, unknown> : {};
      const rawResult = readString(metadata.resultStatus, "Valid");

      return {
        id: activity.id,
        actionLabel:
          activity.action === "PUBLIC_VERIFICATION_SUMMARY_DOWNLOADED"
            ? "Summary downloaded"
            : "Record viewed",
        verifierName: readString(details.name),
        email: readString(details.email),
        phone: readString(details.phone, ""),
        organization: readString(details.organization),
        designation: readString(details.designation, ""),
        purpose: readString(details.purpose),
        notes: readString(details.notes, ""),
        result: rawResult === "Revoked" ? "Revoked" : "Valid",
        ipAddress: readString(metadata.ipAddress),
        userAgent: readString(metadata.userAgent),
        createdAt: activity.created_at
      };
    });
}

function ActivityMetadata({ metadata }: { metadata: Record<string, unknown> | null }) {
  if (!metadata) return null;

  const verifier = metadata.verifier;
  if (typeof verifier !== "object" || verifier === null) return null;

  const details = verifier as Record<string, unknown>;
  const lines = [
    ["Verifier", details.name],
    ["Organization", details.organization],
    ["Email", details.email],
    ["Purpose", details.purpose],
    ["Result", metadata.resultStatus],
    ["IP", metadata.ipAddress]
  ].filter(([, value]) => typeof value === "string" && value.length > 0) as Array<[string, string]>;

  if (lines.length === 0) return null;

  return (
    <div className="mt-2 rounded-lg border border-border bg-muted/40 p-3 text-xs text-muted-foreground">
      {lines.map(([label, value]) => (
        <div key={label} className="flex justify-between gap-3">
          <span>{label}</span>
          <span className="text-right font-medium text-foreground">{value}</span>
        </div>
      ))}
    </div>
  );
}

function Info({ label, value }: { label: string; value: string }) {
  return (
    <div>
      <p className="text-xs uppercase text-muted-foreground">{label}</p>
      <p className="text-sm font-medium">{value}</p>
    </div>
  );
}
